### PR TITLE
Add an 'indented' method to SyntaxProtocol

### PR DESCRIPTION
**Pull request by j-f1**
This method allows for easily indenting a node tree without needing to manually find where to place the new trivia.

I also chose not to implement a version that automatically detects an amount to indent by using `BasicFormat.inferIndentation` since that would probably produce better results if called higher up in the tree, and the user of this API can more easily cache that calculated indentation amount and pass it to `indent` multiple times as needed.

### CODE REVIEW

**Comment by j-f1** [comment]
@swift-ci please test

**Review by ahoppen — COMMENTED** [review]
Sorry for taking sooooooo long to get back to this. We already have `Indenter` in `SwiftSyntaxBuilder`, which serves the same purpose. I don’t think that `Indenter` was ever intended to be public API, but now it’s out there.

I think your API is the nicer one (doesn’t expose a `SyntaxRewriter`) but I think it would be good to unify these two implementations. Could you:
- Mark the `Indenter` class in `SwiftSyntaxBuilder` as deprecated? 
- Migrate all uses of `Indenter` inside swift-syntax to your indentation method?
- Add release note entries for the deprecation of the `Indenter` and the introduction of the new `indented` method?

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
Why might the `indented(by:)` method not be able to be found? Things work fine for me locally.

```
/Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift-syntax/Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:85:27: error: value of type 'Node' has no member 'indented'
      indentedNode = node.indented(by: lastIndentation)
                     ~~~~ ^~~~~~~~
```

**Comment by CodaFi** [comment]
A `Node` is not a kind of Syntax, it’s meta-information about the declaration of a syntax node the builder will spit out. You may want to revert your change in that file.

**Comment by j-f1** [comment]
It’s a local generic type (which must be shadowing the real `Node` type): `func appendInterpolation<Node: SyntaxProtocol>`.

**Comment by ahoppen** [comment]
> Why might the `indented(by:)` method not be able to be found? Things work fine for me locally.
> 
> ```
> /Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift-syntax/Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:85:27: error: value of type 'Node' has no member 'indented'
>       indentedNode = node.indented(by: lastIndentation)
>                      ~~~~ ^~~~~~~~
> ```

You need to add `Indenter.swift` to CMakeLists.txt so we can find it when building swift-syntax using CMake for the compiler.

**Review by ahoppen — COMMENTED** [review]
Thanks you for coming back to this after I took a while, @j-f1. Looks good to me, just a few minor comments.

File: Sources/SwiftBasicFormat/Indenter.swift [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+private class Indenter: SyntaxRewriter {
+  private let indentation: Trivia
+  private var shouldIndent: Bool
+
+  init(indentation: Trivia, indentFirstLine: Bool) {
+    self.indentation = indentation
+    self.shouldIndent = indentFirstLine
+  }
+
+  private func indentIfNeeded() -> [TriviaPiece] {
```
[the remark this answers is not in the document]
Comment (j-f1) [reply Sources/SwiftBasicFormat/Indenter.swift:41]: `indentAfterNewlines` does actually perform the indentation (and returns the indented string) but I’ve renamed `indentIfNeeded`.

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
@swift-ci please test

**Comment by j-f1** [comment]
@swift-ci please test Windows

**Pull request merged**